openapi3/helpers.go

```diff
@@ -39,3 +42,140 @@ func Int64Ptr(value int64) *int64 {
 func Uint64Ptr(value uint64) *uint64 {
 	return &value
 }
+
+type componentRef interface {
+	RefString() string
+	RefPath() *url.URL
+	ComponentType() string
+}
+
+// refersToSameDocument returns if the $ref refers to the same document.
+//
+// Documents in different directories will have distinct $ref values that resolve to
+// the same document.
+// For example, consider the 3 files:
+//
+//	/records.yaml
+//	/root.yaml         $ref: records.yaml
+//	/schema/other.yaml $ref: ../records.yaml
+//
+// The records.yaml reference in the 2 latter refers to the same document.
+func refersToSameDocument(o1 componentRef, o2 componentRef) bool {
+	if o1 == nil || o2 == nil {
+		return false
+	}
+
+	r1 := o1.RefPath()
+	r2 := o2.RefPath()
+
+	if r1 == nil || r2 == nil {
+		return false
+	}
+
+	// refURL is relative to the working directory & base spec file.
+	return r1.String() == r2.String()
+}
+
+// referencesRootDocument returns if the $ref points to the root document of the OpenAPI spec.
+//
+// If the document has no location, perhaps loaded from data in memory, it always returns false.
+func referencesRootDocument(doc *T, ref componentRef) bool {
+	if doc.url == nil || ref == nil {
+		return false
+	}
+
+	refURL := *ref.RefPath()
+	refURL.Path, _, _ = strings.Cut(refURL.Path, "#") // remove the document element reference
+
+	// Check referenced element was in the root document.
+	return doc.url.String() == refURL.String()
+}
+
+// MatchesComponentInRootDocument returns if the given component is identical
+// to a component defined in the root document's '#/components/<type>'.
+// It returns a reference to the schema in the form
+// '#/components/<type>/NameXXX'
+//
+// Of course given it a component from the root document will always match.
+//
+// https://swagger.io/docs/specification/using-ref/#syntax
```

Comment:
Also replace your links to swagger.io with ones to the actual spec at https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md